byterun/win32.c

```diff
@@ -79,17 +84,96 @@ static void caml_win32_sys_error(int errnum)
   caml_raise_sys_error(msg);
 }
 
+#define STATIC_BUFFER_SIZE 1024
+int win32_utf8_read(int fd, void * buf, int n) {
+  HANDLE hConsole = (HANDLE)_get_osfhandle(fd);
+  WCHAR szBuffer[STATIC_BUFFER_SIZE];
+  int real_n = (n > STATIC_BUFFER_SIZE << 2 ? STATIC_BUFFER_SIZE : n >> 2);
+  DWORD nread;
+  if (!ReadConsole(hConsole, szBuffer, real_n, &nread, NULL))
+    return -1;
+  return WideCharToMultiByte(CP_UTF8, 0, szBuffer, nread, buf, n, NULL, NULL);
+}
+
+int win32_utf8_write(int fd, void * buf, int n) {
+  HANDLE hConsole = (HANDLE)_get_osfhandle(fd);
+  int length;
+  WCHAR szBuffer[STATIC_BUFFER_SIZE];
+  LPWSTR buffer;
+  DWORD dwOffset = 0;
+  DWORD dwWritten;
+
+  length = MultiByteToWideChar(CP_UTF8, MB_ERR_INVALID_CHARS, buf, n, NULL, 0);
+  if (length == 0 || length == n) {
+    /* Either invalid UTF-8 or ASCII */
+    return write(fd, buf, n);
+  } else {
+    if (length > STATIC_BUFFER_SIZE) {
+      /* Just resort to write if for some weird reason we run out of memory */
+      if (!(buffer = (LPWSTR)malloc(length * sizeof(WCHAR))))
+        return write(fd, buf, n);
+    } else {
+      buffer = szBuffer;
+    }
+    /* This call must succeed */
+    MultiByteToWideChar(CP_UTF8, 0, buf, n, buffer, length);
+    while (length > 0) {
+      /* Discarding errors at this point */
+      WriteConsole(hConsole, buffer + dwOffset, length, &dwWritten, NULL);
+      length -= dwWritten;
+      dwOffset += dwWritten;
+    }
+    if (length > STATIC_BUFFER_SIZE)
+      free(buffer);
+    return n;
+  }
+}
+#undef STATIC_BUFFER_SIZE
+
+int console_supports_unicode (void) {
+  HANDLE hConsole;
+  CONSOLE_FONT_INFOEX fontInfo;
+  int result = 0;
+  fontInfo.cbSize = sizeof(fontInfo);
+  hConsole =
+    CreateFile(L"CONOUT$", GENERIC_READ | GENERIC_WRITE, FILE_SHARE_WRITE, NULL,
+               OPEN_EXISTING, 0, NULL);
+  if (hConsole != INVALID_HANDLE_VALUE) {
+    if (GetCurrentConsoleFontEx(hConsole, FALSE, &fontInfo))
+      result = ((fontInfo.FontFamily & TMPF_TRUETYPE) != 0);
+    CloseHandle(hConsole);
+  }
+
+  return result;
+}
+
 int caml_read_fd(int fd, int flags, void * buf, int n)
 {
   int retcode;
   if ((flags & CHANNEL_FLAG_FROM_SOCKET) == 0) {
     caml_enter_blocking_section();
-    retcode = read(fd, buf, n);
+#if WINDOWS_UNICODE
+    if ((flags & CHANNEL_FLAG_CONSOLE) && console_supports_unicode()) {
+#else
+    if ((flags & CHANNEL_FLAG_CONSOLE) && GetConsoleCP() == CP_UTF8
+        && console_supports_unicode()) {
+#endif
+      /* Cannot perform a UTF-8 read unless the buffer is at least 4 bytes */
```

Comment:
Is there a risk that some applications would be written to read bytes one at a time (so with `n == 1` always), and happen to be run on unicode-supporting systems? Would it be possible to fallback to just a `read` call in that case (I guess that means designed-for-unicode code could sometimes behave incorrectly?)?

Reply:
Falling back to `read` would mean that you might get garbage (if you're lucky - often it's a segfault). I wasn't certain, but it seemed unlikely that you'd ever use this in C stubs - it's largely used internally in a context where it's refilling the channel's buffer.

I'm not sure whether the support of these two is worth extending to Unix.read and Unix.write (they will both exhibit the same problems if you use Unix.stdin and Unix.stdout/Unix.stderr)